command/operator_migrate.go

```diff
@@ -0,0 +1,287 @@
+
+	f.StringVar(&StringVar{
+		Name:   "start",
+		Target: &c.flagStart,
+		Usage:  "Only copy keys lexicographically at or after this value.",
+	})
+
+	f.BoolVar(&BoolVar{
+		Name:   "reset",
+		Target: &c.flagReset,
+		Usage:  "Reset the migration lock. No migration will occur.",
+	})
+
+	return set
+}
+
+func (c *OperatorMigrateCommand) AutocompleteArgs() complete.Predictor {
+	return nil
+}
+
+func (c *OperatorMigrateCommand) AutocompleteFlags() complete.Flags {
+	return c.Flags().Completions()
+}
+
+func (c *OperatorMigrateCommand) Run(args []string) int {
+	c.logger = logging.NewVaultLogger(log.Info)
+	f := c.Flags()
+
+	if err := f.Parse(args); err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	if c.flagConfig == "" {
+		c.UI.Error("Must specify exactly one config path using -config")
+		return 1
+	}
+
+	config, err := c.loadMigratorConfig(c.flagConfig)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error loading configuration from %s: %s", c.flagConfig, err))
+		return 1
+	}
+
+	if err := c.migrate(config); err != nil {
+		c.UI.Error(fmt.Sprintf("Error migrating: %s", err))
+		return 2
+	}
+
+	if c.flagReset {
+		c.UI.Output("Success! Migration lock reset (if it was set).")
+	} else {
+		c.UI.Output("Success! All of the keys have been migrated.")
+	}
+
+	return 0
+}
+
+// migrate attempts to instantiate the source and destinations backends,
+// and then invoke the migration the the root of the keyspace.
+func (c *OperatorMigrateCommand) migrate(config *migratorConfig) error {
+	from, err := c.newBackend(config.StorageSource.Type, config.StorageSource.Config)
+	if err != nil {
+		return errwrap.Wrapf("error mounting 'storage_source': {{err}}", err)
+	}
+
+	if c.flagReset {
+		if err := SetMigration(from, false); err != nil {
+			return errwrap.Wrapf("error reseting migration lock: {{err}}", err)
+		}
+		return nil
+	}
+
+	to, err := c.newBackend(config.StorageDestination.Type, config.StorageDestination.Config)
+	if err != nil {
+		return errwrap.Wrapf("error mounting 'storage_destination': {{err}}", err)
+	}
+
+	startTime, err := CheckMigration(from)
+	if err != nil {
+		return errors.New("error checking migration status")
+	}
+
+	if startTime != "" {
+		return fmt.Errorf("Storage migration in progress (started: %s).", startTime)
+	}
+
+	if err := SetMigration(from, true); err != nil {
+		return errwrap.Wrapf("error setting migration lock: {{err}}", err)
+	}
+
+	defer SetMigration(from, false)
+
+	return c.migratePath(context.Background(), "", from, to)
+}
+
+// migratePath copies all keys at at given level in lexicographic order.
+// It will call itself recursively to traverse the entire subtree.
+func (c *OperatorMigrateCommand) migratePath(ctx context.Context, path string, from physical.Backend, to physical.Backend) error {
+	keys, err := from.List(ctx, path)
+	if err != nil {
+		return err
+	}
+
+	sort.Strings(keys)
+	for _, key := range keys {
+		if strings.HasSuffix(key, "/") {
+			err := c.migratePath(ctx, path+key, from, to)
```

Comment:
Rather than make this recursive you could use the `ScanView` function that iteratively hits each key. Could save a lot of call stack memory. 

https://github.com/hashicorp/vault/blob/b7d6d55ac12c7ffbf9f78dbde26acd19ec7662f7/logical/storage.go#L61

Reply:
+1

Reply:
The recursion has been removed, being replaced with a simple depth-first iterative traversal.

`ScanView` wasn't a clean drop-in because it is breadth-first and doesn't guarantee ordering, both of which somewhat complicate the `-start` process.